tests/distributed/test_fsdp_optim_state.py

```diff
@@ -0,0 +1,165 @@
+    torch.manual_seed(0)  # set seed for determinism
+    optim.zero_grad()
+    batch_size = 1024
+    device = ta.lazy_device()
+
+    data1 = torch.rand(batch_size, 1024).to(device)
+    data2 = torch.zeros(batch_size, dtype=torch.int64).to(device)
+    loss = model(data1)
+    loss = torch.nn.functional.nll_loss(loss, data2)
+    loss.backward()
+
+
+def _check_optim_state(fsdp_osd1, fsdp_osd2):
+    state1 = fsdp_osd1['state']
+    state2 = fsdp_osd2['state']
+    for key in state1.keys():
+        dict1 = state1[key]
+        dict2 = state2[key]
+        for state_name in dict1.keys():
+            tensor1 = dict1[state_name]
+            tensor2 = dict2[state_name]
+            assert torch.equal(tensor1, tensor2)
+
+
+def _check_optim_param_groups(fsdp_osd1, fsdp_osd2):
+    param1 = fsdp_osd1['param_groups']
+    param2 = fsdp_osd2['param_groups']
+
+    for (value1, value2) in zip(param1, param2):
+        for key in value1.keys():
+            assert value1[key] == value2[key]
+
+
+class FSDPOptimStateTest(MultiProcessTestBase):
+
+    @property
+    def world_size(self) -> int:
+        return 4
+
+    @skip_if_lt_x_gpu(2)
+    @init_pg("lazy")
+    def test_fsdp_optim_state_flatten_gpu4_4(self):
+        # we first init a model
+        config1 = Config()
+        config1.dist.fsdp.size = self.world_size
+        model_1, optim_1 = _init_model(config=config1)
+        # iter 10 steps
+        _step_model(model_1, optim_1, 10)
+        # get the optim_state_dict for model1
+        fsdp_osd1 = FSDP.optim_state_dict(
+            model_1, optim_1, full_state_dict="FULL_STATE_DICT")
+
+        # init a new model with same world_size
+        config2 = Config()
+        config2.dist.fsdp.size = self.world_size
+        model_2, optim_2 = _init_model(config=config2)
+        # we may change fsdp_osd1 in load_optim_state_dict
+        fsdp_osd1_ = copy.deepcopy(fsdp_osd1)
+        # model_2 load the optim_state_dict from model_1
+        fsdp_osd_to_load = FSDP.load_optim_state_dict(model_2, fsdp_osd1_,
+                                                      optim_2,
+                                                      "FULL_STATE_DICT")
+        optim_2.load_state_dict(fsdp_osd_to_load)
+        _step_model_without_update(model_2, optim_2)
+        fsdp_osd2 = FSDP.optim_state_dict(
+            model_2, optim_2, full_state_dict="FULL_STATE_DICT")
+
+        _check_optim_state(fsdp_osd1, fsdp_osd2)
+        _check_optim_param_groups(fsdp_osd1, fsdp_osd2)
+
+    @skip_if_lt_x_gpu(2)
+    @init_pg("lazy")
+    def test_fsdp_optim_state_flatten_gpu4_2(self):
+        config1 = Config()
+        config1.dist.fsdp.size = self.world_size
+        model_1, optim_1 = _init_model(config=config1)
+        _step_model(model_1, optim_1, 10)
+        fsdp_osd1 = FSDP.optim_state_dict(
+            model_1, optim_1, full_state_dict="FULL_STATE_DICT")
+
+        # we create a new group with world_size / 2
+        new_world_size = self.world_size / 2
```

Comment:
// 2

Reply:
also need check world_size % 2 == 0

Reply:
do we need to check world_size % 2 == 0 if use // 2? does the number of world_size must be even or the size of each sharding group must be equal?

Reply:
yes, must be equal